src/clients/HubPoolClient.ts

```diff
@@ -646,11 +643,43 @@ export class HubPoolClient {
     this.disputedRootBundles.push(
       ...events["RootBundleDisputed"].map((event) => spreadEventWithBlockNumber(event) as DisputedRootBundle)
     );
-    this.executedRootBundles.push(
-      ...events["RootBundleExecuted"]
-        .filter((event) => !this.configOverride.ignoredHubExecutedBundles.includes(event.blockNumber))
-        .map((event) => spreadEventWithBlockNumber(event) as ExecutedRootBundle)
-    );
+
+    const configStoreVersions: { [blockNumber: number]: number } = {};
+    for (const event of events["RootBundleExecuted"]) {
+      if (this.configOverride.ignoredHubExecutedBundles.includes(event.blockNumber)) {
+        continue;
+      }
+
+      if (!isDefined(configStoreVersions[event.blockNumber])) {
+        const timestamp = (await event.getBlock()).timestamp;
+        const version = this.configStoreClient.getConfigStoreVersionForTimestamp(timestamp);
+        configStoreVersions[event.blockNumber] = version;
+      }
+      const version = configStoreVersions[event.blockNumber];
+
+      const executedRootBundle = spreadEventWithBlockNumber(event) as ExecutedRootBundle;
+      const { l1Tokens, runningBalances } = executedRootBundle;
+      const nTokens = l1Tokens.length;
```

Comment:
We're paying a _fairly_ high costs (extra RPCs in the hot path) for padding the incentiveBalances in the case of pre-UBA versions.

Since we already know the nTokens, we should always be able to correctly slice the array, regardless. What value are we getting out of this padding that we couldn't get just by _always padding_ if there aren't enough values? Note: these are executed, so they technically should already be validated, right?

Reply:
Not sure whether I'm understanding your comment correctly, but the challenge is in knowing when the `incentiveBalances` should be zeroes, and when it should be a slice of `runningBalances`. This PR started out by inferring the correct shape of `runningBalances` and `incentiveBalances` based solely on the length of `runningBalances` in the on-chain event. Nick had a good point that we should only make this evaluation based on the version number, which is how it ended here.

Fortunately, based on your other feedback about the cost of querying the blocks for each update, I realised that we already have all of the necessary information, and that we only lacked a helper method in the `ConfigStoreClient`. See here for a PR to add it in: https://github.com/across-protocol/sdk-v2/pull/230.

Reply:
This is great! +1. I was mostly just worried about the RPC calls. If we don't make them, then it's no problem!